peer/pendingheap/heap.go

```diff
@@ -0,0 +1,186 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pendingheap
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+)
+
+type pendingHeap struct {
+	sync.RWMutex
+
+	transport peer.Transport
+	peers     []*peerScore
+
+	// next is an incrementing counter for every push, which is compared when
+	// scores are equal. This ends up implementing round-robin when scores are
+	// equal.
+	next int
+}
+
+func (ph *pendingHeap) Choose(ctx context.Context, req *transport.Request) peer.StatusPeer {
+	ph.Lock()
+	ps, ok := ph.popPeer()
+	if !ok {
+		ph.Unlock()
+		return nil
+	}
+
+	// Note: We push the peer back to reset the "next" counter.
+	// This gives us round-robin behavior.
+	ph.pushPeer(ps)
+
+	ph.Unlock()
+	return ps.peer
+}
+
+func (ph *pendingHeap) Add(p peer.StatusPeer) peer.Subscriber {
+	if p == nil {
+		return nil
+	}
+
+	ps := &peerScore{peer: p, heap: ph}
+	ps.score = scorePeer(p)
+
+	ph.Lock()
+	ph.pushPeer(ps)
+	ph.Unlock()
+	return ps
+}
+
+func (ph *pendingHeap) Remove(p peer.StatusPeer, sub peer.Subscriber) {
+	ps, ok := sub.(*peerScore)
+	if !ok {
+		return
+	}
+
+	ph.Lock()
+	ph.delete(ps.idx)
+	ph.Unlock()
+}
+
+func (ph *pendingHeap) notifyStatusChanged(ps *peerScore) {
+	status := ps.peer.Status()
+	ph.Lock()
+	ps.status = status
+	ps.score = scorePeer(ps.peer)
+	ph.update(ps.idx)
+	ph.Unlock()
+}
+
+func (ph *pendingHeap) Len() int {
+	l := len(ph.peers)
+	return l
+}
+
+// Less returns whether the left peer has a lower score. If the scores are
+// equal, it returns the older peer (where "last" is lower.)
+func (ph *pendingHeap) Less(i, j int) bool {
+	p1 := ph.peers[i]
+	p2 := ph.peers[j]
+	if p1.score == p2.score {
+		return p1.last < p2.last
+	}
+	return p1.score < p2.score
+}
+
+// Swap implements the heap.Interface. Do NOT use this method directly.
+func (ph *pendingHeap) Swap(i, j int) {
```

Comment:
Don't `Swap`, `Pop`, `Push` and `delete` need locks?

Reply:
Documented. These are here to implement the heap.Interface, which we call indirectly through heap.Push and heap.Pop. They must be called in the context of a lock higher in the stack.

✔️